### PR TITLE
Publish and consume MLIR artifact for Windows CI

### LINKED ISSUE #776: Long build time for Windows CI
I think adding the WindowCI as mandatory is a good move but at this time:
1) the build takes 2:20 hours to build (last one that passed)
2) we don't have a way to test it (IBM team has access to Mac/Linux x86 but not Windows). Would it be possible to make a virtual machine available to our team for us to develop when there are errors specific to Windows.

My personal preference would be to make the Windows CI mandatory once these two points are addressed.

### PR DESCRIPTION
**Pull request by NathanielMcVicar**
Previously, Windows CI was much slower than other platforms because MLIR was rebuilt for every run using. This change brings Windows CI into line with other pipelines by storing MLIR after building from a new LLVM commit.

closes #776

### CODE REVIEW

File: .azure-pipelines/Windows-CI.yml [hunk 1]
```diff
@@ -43,11 +44,49 @@ jobs:
...
       call "%ProgramFiles(x86)%\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
       call onnx-mlir\utils\install-mlir.cmd
     displayName: Install mlir
     workingDirectory: $(Agent.BuildDirectory)
+    condition: and(succeeded(), eq(variables['MLIRBuildID'], 'true'))
+
+  - publish: $(Agent.BuildDirectory)/llvm-project/build/install
+    artifact: $(MLIRArtifactName)
+    displayName: Publish MLIR Artifact
+    condition: and(succeeded(), eq(variables['MLIRBuildID'], 'true'))
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:77]: MLIRBuildID? Should this be BuildMLIR?

File: .azure-pipelines/Windows-CI.yml [hunk 2]
```diff
@@ -43,11 +44,49 @@ jobs:
...
+    displayName: Check for MLIR Artifact
+    failOnStderr: true
+    workingDirectory: $(Agent.BuildDirectory)
+
   - script: |
       call "%ProgramFiles(x86)%\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
       call onnx-mlir\utils\install-mlir.cmd
     displayName: Install mlir
     workingDirectory: $(Agent.BuildDirectory)
+    condition: and(succeeded(), eq(variables['MLIRBuildID'], 'true'))
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:72]: Should this be BuildMLIR?

File: .azure-pipelines/Windows-CI.yml [hunk 3]
```diff
@@ -43,11 +44,49 @@ jobs:
...
+    inputs:
+      source: specific
+      project: onnx
+      pipeline: 9
+      runVersion: specific
+      runId: $(MLIRBuildID)
+      artifact: $(MLIRArtifactName)
+      path: $(Agent.BuildDirectory)\llvm-project\build
+    displayName: Download MLIR Artifact
+    condition: and(succeeded(), neq(variables['MLIRBuildID'], 'false'))
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:89]: Here too. You have created both `BuildMLIR` and `MLIRBuildID` but then are only using one of them, can we get away with one?

**Review by sstamenova — CHANGES_REQUESTED** [review]

**Review by sstamenova — APPROVED** [review]
LGTM

File: .azure-pipelines/Windows-CI.yml [hunk 4]
```diff
@@ -43,11 +44,48 @@ jobs:
...
+    inputs:
+      source: specific
+      project: onnx
+      pipeline: 9
+      runVersion: specific
+      runId: $(MLIRBuildID)
+      artifact: $(MLIRArtifactName)
+      path: $(Agent.BuildDirectory)\llvm-project\build
+    displayName: Download mlir artifact
+    condition: and(succeeded(), neq(variables['MLIRBuildID'], 'NotFound'))
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:88]: ne not neq

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@NathanielMcVicar Thanks Thanks Thanks

**Comment by NathanielMcVicar** [comment]
@AlexandreEichenberger We're resolving some permission issues with onnx-pipelines on our end, but I have a plan for a slightly slower workaround if we can't do it by early next week.

File: .azure-pipelines/Windows-CI.yml [hunk 5]
```diff
@@ -48,17 +48,34 @@ jobs:
       $mlir_match = Select-String -Path onnx-mlir\utils\clone-mlir.sh -Pattern "git checkout ([\da-f]+)"
       $mlir_tag = "MLIR_Windows_$($mlir_match.Matches.Groups[1].Value)"
 
-      $build_list = az pipelines build list --org https://dev.azure.com/onnx-pipelines --project onnx --definition-ids 9 --tags $mlir_tag --status completed --result succeeded | ConvertFrom-Json
-      $most_recent = $build_list | Sort-Object startTime | Select-Object -Last 1
-      $build_id = $most_recent.id
+      # Use this version to select artifact by tag, once permissions are enabled this will be significantly faster
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:51]: Which parts would have to be removed to make this section work? Everything beside it?

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

File: .azure-pipelines/Windows-CI.yml (hunk 5 above)
Comment (NathanielMcVicar) [reply .azure-pipelines/Windows-CI.yml:51]: That's a good point, I'll clarify in the comment

**Comment by NathanielMcVicar** [comment]
@AlexandreEichenberger I pushed again with a better comment to test the caching. If this works, we're good to go.

**Comment by tungld** [comment]
@jenkins-droid  test this please

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by gongsu832** [comment]
@jenkins-droid test this please

**Comment by NathanielMcVicar** [comment]
This should be ready to merge now @AlexandreEichenberger @gongsu832. It looks like it works in both cases (building and consuming the artifact). I'm still working internally to get the storage issue resolved, but I'm not sure when the relevant people will be back in the office. It shouldn't be blocking.

File: .azure-pipelines/Windows-CI.yml [hunk 6]
```diff
@@ -40,18 +50,78 @@ jobs:
   - script: |
       call "%ProgramFiles(x86)%\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
       call onnx-mlir\utils\install-protobuf.cmd
-    displayName: Install protobuf
+    displayName: Install lit and protobuf
```
Comment (sstamenova) [inline .azure-pipelines/Windows-CI.yml:53]: Is the `displayName` wrong now?

**Comment by gongsu832** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@NathanielMcVicar Thank you for the fix. I presume it goes 2x faster, is that right! Looking forward to many faster turnaround time now that we can go through verification much faster. Big thanks

**Comment by NathanielMcVicar** [comment]
> I presume it goes 2x faster

@AlexandreEichenberger It looks like times are down from 2 hours+ to 35 minutes. I think we could cache protobuf and speed up the scripts to maybe get to 30 minutes, but it doesn't seem worth the complexity. Regardless, happy I could help!

**Pull request merged**